src/gurt/misc.c

```diff
@@ -566,3 +567,87 @@ void d_getenv_int(const char *env, unsigned *int_val)
 	D_DEBUG(DB_TRACE, "d_getenv_int(), get ENV %s as %d.\n", env, value);
 	*int_val = value;
 }
+
+/**
+ * Write formatted data to d_string_buffer_t
+ *
+ * Create a string with the same text that would be created by printf(). The
+ * result string is stored as a C string in a buffer pointed by buf. The
+ * d_string_buffer_t internal buffer grows as text is written.
+ *
+ * \param[in,out] buf    string object where the formatted string will be
+ *                       stored. Subsequent write operations to the same string
+ *                       object will be append at the end of the buffer.
+ * \param[in]     format this is the string that contains the text to be
+ *                       written to d_string_buffer_t. Please refer to the
+ *                       printf() documentation for full details of how to use
+ *                       the format tags.
+ * \return               0 on success, errno code on failure.
+ */
+int
+d_write_string_buffer(struct d_string_buffer_t *buf, const char *format, ...)
+{
+	int n;
+	int size = 64;
+	char *new_buf;
+	va_list ap;
+
+	if (buf == NULL || buf->status != 0) {
+		return -DER_NO_PERM;
+	}
+
+	if (buf->str == NULL) {
+		D_ALLOC(buf->str, size);
+		if (buf->str == NULL) {
+			buf->status = -DER_NOMEM;
+			return -DER_NOMEM;
+		}
+		buf->str_size = 0;
+		buf->buf_size = size;
+	}
+
+	while (1) {
+		va_start(ap, format);
+		size = buf->buf_size - buf->str_size;
+		n = vsnprintf(buf->str + buf->str_size, size, format, ap);
+		va_end(ap);
+
+		if (n < 0) {
+			buf->status = -DER_TRUNC;
+			return -DER_TRUNC;
+		}
+
+		if ((buf->str_size + n) < buf->buf_size) {
+			buf->str_size += n;
```

Comment:
should buf->status be set here as well? if you ever encounter any error, buf->status will not get reset after retry even if everything was successful

Reply:
Not really, once the `d_string_buffer_t` is dirty (an error happened) the only way to clean it is by calling `d_free_string` For the sake of consistency, I just added `buf->status = DER_SUCCESS;`